### DISCUSSION TITLE
How do I add a clickable area with an external link?

### DISCUSSION THREAD

**Discussion by steinerlein** [pymupdf/PyMuPDF#1373] [Looking for help] [answered]
Hi, I'm looking to add clickable links to existing PDFs.
I can already retrieve the location of corresponding Text and I can also construct the file destination.
However, it is unclear to me how to actually insert a clickable area.

Here's what I have so far:
``` import fitz
import re

doc = fitz.open("RST-004285_00.pdf")

page = doc[0]
blocks = page.get_text("blocks")
blockRects = []
for block in blocks:
    stringPart = block[4]
    parts = stringPart.split("\n")
    try:
        if re.search("rst", parts[2].lower()):
            print(parts[2]+"-"+parts[3])
            blockRects.append(fitz.Rect(block[0], block[1], block[2], block[3]))
    except IndexError as e:
        pass

for block in blockRects:
    page.add_squiggly_annot(block)
    blockLink = fitz.Link
    blockLink.rect = block
    blockLink.uri = "C:/Users/Farid/Desktop/test.pdf"
    blockLink.is_external = True
    page.insert_link(blockLink)
    
doc.save("a-squiggly.pdf")
```

I understand handling links is a new feature and if it's not possible to do what I want, please let me know :)

Thanks!

**Comment by JorjMcKie** [comment, marked as answer]
> I understand handling links is a new feature and if it's not possible to do what I want, please let me know :)

On the contrary: it is quite an old feature 😉and you sure can do it.

Some background:
Links and annotations used to be supported for all document types in MuPDF. Annotation support has then been restricted to PDFs at some point.
In **_PyMuPDF_**, annotations have been added much later than links. This is why they are handled via a differently looking API.

There is `page.insert_link(link)` to insert a new link on a page. The argument must a dictionary as described [here](https://pymupdf.readthedocs.io/en/latest/page.html#reading-pertains-to-method-get-links-and-the-first-link-property-chain), look at section heading "Description of get_links() Entries".
